core/payment/src/processor.rs

```diff
@@ -179,30 +180,64 @@ impl PaymentProcessor {
         }
         // TODO: Sender should be included in transaction details and checked as well
 
+        // Verify agreement payments
         let agreement_dao: AgreementDao = self.db_executor.as_dao();
-        let agreement = agreement_dao.get(agreement_id.clone(), payee_id).await?;
-        match agreement {
-            None => {
-                let msg = format!("Agreement not found: {}", agreement_id);
-                return Err(PaymentError::Verification(msg).into());
+        for agreement_payment in payment.agreement_payments.iter() {
+            let agreement_id = agreement_payment.agreement_id.clone();
+            let agreement = agreement_dao.get(agreement_id.clone(), payee_id).await?;
+            match agreement {
+                None => {
+                    let msg = format!("Agreement not found: {}", agreement_id);
+                    return Err(PaymentError::Verification(msg).into());
+                }
+                Some(agreement) if &agreement.payee_addr != &payee_addr => {
+                    let msg = format!(
+                        "Invalid payee address for agreement {}. {} != {}",
+                        agreement_id, &agreement.payee_addr, &payee_addr
+                    );
+                    return Err(PaymentError::Verification(msg).into());
+                }
+                Some(agreement) if &agreement.payer_addr != &payer_addr => {
+                    let msg = format!(
+                        "Invalid payer address for agreement {}. {} != {}",
+                        agreement_id, &agreement.payer_addr, &payer_addr
+                    );
+                    return Err(PaymentError::Verification(msg).into());
+                }
+                _ => (),
             }
-            Some(agreement) if &agreement.payee_addr != &payee_addr => {
-                let msg = format!(
-                    "Invalid payee address. {} != {}",
-                    &agreement.payee_addr, &payee_addr
-                );
-                return Err(PaymentError::Verification(msg).into());
-            }
-            Some(agreement) if &agreement.payer_addr != &payer_addr => {
-                let msg = format!(
-                    "Invalid payer address. {} != {}",
-                    &agreement.payer_addr, &payer_addr
-                );
-                return Err(PaymentError::Verification(msg).into());
+        }
+
+        // Verify activity payments
```

Comment:
Do you have any particular idea on how to do this? I see two options and both seem worse than repeating similar code:
* Use macro => messy & obscure :face_with_head_bandage: 
* Introduce a trait that is implemented by both `Activity` and `Agreement` and create a template method that accepts objects implementing that trait => clean but very verbose (it might turn out to be more code)

Reply:
I was thinking about the second option but I agree with you. So your choice ;)

Reply:
I'll leave for now. Doing extra work to have more code doing the same thing seems kind of pointless to me.

Reply:
Ok